ping/ping.c

```diff
@@ -555,6 +555,23 @@ main(int argc, char **argv)
 	int target_ai_family = hints.ai_family;
 	hints.ai_family = AF_UNSPEC;
 
+	/*
+	 * Allow to use -I<ifname|scope-id> for IPv6 link-local scope address
+	 * when passing it to getaddrinfo():
+	 * i.e. -I<ifname|scope-id> => %<ifname|scope-id>
+	 */
+	unsigned char buf[sizeof(struct in6_addr)];
+	if (!strchr(target, '%') && inet_pton(AF_INET6, target, buf) > 0 &&
+		(IN6_IS_ADDR_LINKLOCAL(buf) || IN6_IS_ADDR_MC_LINKLOCAL(buf))) {
+		if (rts.device) {
```

Comment:
@jonharrisonms @gobenji @nmeyerhans I haven't tested this with VRF (9e08707d743b29e853df81bd7def1729e3afe55d, 2705c8248281fbb8efaa5326ab1d0ed0a670bd3d).

It looks like basic getopts parsing for device works at least for `-I <source_addr> -I <ifname>`:
https://github.com/iputils/iputils/blob/854873bdd28fcdd9cc3fe0c2d29c083a07d07a86/ping/ping.c#L388-L413

because all of these works (`2620:113:80c0:8000:10:161:28:49` is local IPv6 global scope dynamic noprefixroute, `fe80::5054:ff:fe55:be42` is remote IPv6 link scope) - verified with `tcpdump`:
```
ping fe80::5054:ff:fe55:be42%eth0 -I 2620:113:80c0:8000:10:161:28:49
ping fe80::5054:ff:fe55:be42 -I eth0 -I 2620:113:80c0:8000:10:161:28:49
ping fe80::5054:ff:fe55:be42 -I 2620:113:80c0:8000:10:161:28:49 -I eth0
```
(also verified it works with this change).

Also maybe I should add here also check whether `rts.device` and not apply it when invalid ifname, not sure if new functionality is not misleading:
```
$ ping fe80::5054:ff:fe55:be42 -I noexisting # original
ping: unknown iface: noexisting

$ ./builddir/ping/ping fe80::5054:ff:fe55:be42 -I noexisting # with this change
./builddir/ping/ping: fe80::5054:ff:fe55:be42%noexisting: Name or service not known
```